documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md

```diff
@@ -0,0 +1,830 @@
+costs directly. However, the complexity of sensitive text designation exists no
+matter where it goes in our compute pipeline. In the indexer workers, it will
+naturally extend the time each worker spends on any given document. As such, the
+overall amount of time the indexer workers are running will increase, and so
+will their cost to us. Elsewhere, for example, in the Airflow local executor, we
+would pay for it in decreased resources available for other tasks, extending the
+time of those tasks, and potentially decreasing the scheduler's stability (if
+resource usage increases a lot).
+
+### Tools & packages
+
+<!-- Describe any tools or packages which this work might depend on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+We do not require any new tools or packages to implement this change.
+
+### Other projects or work
+
+<!-- Note any projects this plan depends on. -->
+
+As mentioned in [the project links](#project-links), this plan assumes
+[the removal of the ingestion server](https://github.com/WordPress/openverse/issues/3925)
+is finished by the time this plan makes modifications to the data refresh and
+indexer worker.
+
+## Alternatives
+
+<!-- Describe any alternatives considered and why they were not chosen or recommended. -->
+
+### Elasticsearch ingest pipeline
+
+I spent a few days experimenting with and trying to see whether an ingest
+pipeline would be a viable way to approach the creation of the sensitivity list
+for this project. I believe it is generally viable, and I was able to make it
+work locally with our local sample data and the full sensitive terms list. I
+wrote about my evaluation of the ingest pipeline approach in these three
+comments on the implementation plan issue:
+
+1. [Describing the approach in general and initial testing](https://github.com/WordPress/openverse/issues/3336#issuecomment-2316894501)
+1. [Further local testing with the complete sensitive terms list](https://github.com/WordPress/openverse/issues/3336#issuecomment-2325659270)
+1. [Testing the ingest pipeline in staging](https://github.com/WordPress/openverse/issues/3336#issuecomment-2325684604)
+
+Ultimately, despite ingest pipelines working fine to create the sensitivity
+designations list, I don't think it's the right approach for us. As noted in the
+final comment, whilst dedicated ingest nodes are not required to work with
+ingest pipelines, a cluster still needs at least one node with the ingest role
+to run ingest pipelines (it just isn't required to be those nodes' only role).
+In order to service that requirement, we would need to introduce a new
+complexity into our Elasticsearch cluster. We do not have the capacity (time and
+knowledge) to fulfil this requirement. Therefore, we should not use
+Elasticsearch ingest pipelines to generate the sensitivity designation.
+
+### Move sensitive text detection earlier in the data refresh pipeline
+
```

Comment:
I did consider that, and for all the money in the world I would have sworn I'd written about it in the IP, but I have been thinking of [this comment of mine from the issue](https://github.com/WordPress/openverse/issues/3336#issuecomment-2288432417). It definitely warrants covering in the IP directly.

The things I like about it are the ones you shared, and that we can avoid needlessly re-running the check on works.

The things that make me nervous about it, which made me not want to pursue it, are:

1. A lack of consistency guaranteed in a single data refresh cycle if we are updating sensitivity. If a data refresh runs during the update, then it will have records with indications from the previous list and others from the new list. We could block data refreshes from occurring if a sensitivity update needs to happen (when we change the terms list) to prevent that. Assuming it's possible to do a full sensitivity update in <24 hours (assumed because our current processes are able to accomplish that) then it isn't that big a deal, it would delay the data refresh by that amount of time.
2. Uncertainty about whether we would need two different implementations of sensitive text detection. Our ingestion happens in Python, but batched updates in SQL. As far as I know, we don't have good ways to write parts of ingestion in SQL, and no way to write any part of batched updates in Python (due to RDS's lack of support for ulpython). So my assumption is that we would need a version in SQL to account for the batched update version, and then another in Python for ingestion of new records. However, it occurred to me now that we could use a PGSQL function for this, and then use an insert trigger to derive the `sensitive_text` sensitivity, and the same function could be referenced in the batched update. In other words, I think it may be possible to create a single implementation that accounts for both.
  - My primary concern with two separate implementations is in one not being used often enough for us to be sure that it's "still working". We have not made many changes to the sensitive terms list (I think maybe once, if ever, since implementing its usage). If the SQL version used for the batched update doesn't get regularly used, we risk an unfortunate surprise of it being broken when we make a change to the terms. To fully test this locally, I think we'd need to use the full terms list, or else try to identify all the edge cases (emoji, unicode, special characters, etc), which will need special handling/escaping for put into SQL. Not exercising that regularly is a big risk, long term. The consequence is the surprise of it not working when we most expect and rely on it to work.
3. Uncertainty about the performance impact on the scheduler which is also our executor, due to local executor usage. This is only relevant if the implementation is in Python, which as I've written in the previous point, may not be necessary. So, this is less of a concern to me, as I feel comfortable offloading this kind of work to PG.

I'm actually curious to look more into this. @stacimc what do you think about these points? Do they fit with your understanding of the situation around a popularity-refresh-like approach, _viz_ the limitations and approach we would need to take? In particular, do you have an opinion or sense about point 2, and whether we would be able to have a single implementation rather than a SQL and Python one separately? A single implementation and just accepting a delay to one data refresh using our existing concurrency controls would resolve the questions I had about it. In which case, I think it's undeniably the better approach.

Reply:
To point 1, exactly, it would be very easy to add the `recalculate_sensitivity` DAG to the list of blocking DAGs for the data refresh, and that should be fine _as long as recalculating sensitivity doesn't take too long_ like you said. We'd expect this to happen infrequently, as well.

To point 2, we actually can do SQL updates during ingestion by adding a [CalculatedColumn](https://teamopenverse.wordpress.com/2024/07/25/elasticache-service-update/); this column is not populated by the provider scripts, but given a SQL function that runs during the [upsert query](https://github.com/WordPress/openverse/blob/a4b1bf4e121373d98a2a81bc38a4556a86c05695/catalog/dags/common/loader/sql.py#L316). This is what we do with standardized popularity; the provider ingesters don't "know about" popularity at all. Then the same SQL function is used for the batched popularity updates.

My biggest question is about performance as well, because a full batched update of all records is quite time consuming. Instead I think we should SELECT only the records that _need_ to be updated after an update to the terms list -- ie, all records that match any of the sensitive terms that were either deleted or added. I'm not sure how performant _that_ query would be, that would be interesting to look into.

Reply:
> To point 2, we actually can do SQL updates during ingestion by adding a [CalculatedColumn](https://teamopenverse.wordpress.com/2024/07/25/elasticache-service-update/); this column is not populated by the provider scripts, but given a SQL function that runs during the [upsert query](https://github.com/WordPress/openverse/blob/a4b1bf4e121373d98a2a81bc38a4556a86c05695/catalog/dags/common/loader/sql.py#L316). This is what we do with standardized popularity; the provider ingesters don't "know about" popularity at all. Then the same SQL function is used for the batched popularity updates.

Perfect!

> My biggest question is about performance as well, because a full batched update of all records is quite time consuming. Instead I think we should SELECT only the records that need to be updated after an update to the terms list -- ie, all records that match any of the sensitive terms that were either deleted or added. I'm not sure how performant that query would be, that would be interesting to look into.

Just to make sure I understand your suggestion correctly, are you saying that we would use the `where_query` to `SELECT * FROM image WHERE check_for_sensitive_terms((title, description, tags->name), <new terms>) OR check_for_sensitive_terms(description, <new terms>) OR check_for_sensitive_terms(tags->name, <new terms>)` (where `check_for_sensitive_terms` is the function used in all sensitive term checks including on insert)?

<small>As a bit of an aside, this is actually an interesting question to me that I feel I don't yet understand. Is it an optimisation to filter the set of records beforehand, rather than doing a full update of the table? To my mind, the number of invocations of `check_for_sensitive_terms` would be identical in each, so it comes down to whether PG is happier doing that as part of a full table scan in the select rather than in the individual update batches. Either way, the performance _viz_ the check is O(n) where `n` is the number of rows. If there is a different way to view this, it would be helpful for me to incorporate into my understanding of how to approach batched updates where the update is conditional on a derived quantity (so not simple equality, for example).</small>

Generally, though, it should be easy to test this. For one, we'd need to run this same select at least one time to update based on the current sensitive terms list. We can verify the viability of the `select` pretty easily, in that case, by manually running that query with a constructed sensitive terms list.

<details>
<summary>This function almost works to do it using PL/pgSQL... the terms list is just "water" and "running", but I've tested it with the regular terms list and it falls over on emojis and things like the <code>@</code> symbol.</summary>

```sql
CREATE OR REPLACE FUNCTION check_sensitive_terms(fields text[]) RETURNS BOOLEAN AS $$
  DECLARE
    sensitive_terms_list text;
    sensitive_terms_regexps text[];
    found_match boolean;
    term text;
    term_regexp text;
    field text;
  BEGIN
    sensitive_terms_list := encode(decode(
      'CndhdGVyCnJ1bm5pbmcKCg==',
      'base64'
    ), 'escape');

    sensitive_terms_regexps := ARRAY[]::text[];

    FOREACH term IN ARRAY regexp_split_to_array(sensitive_terms_list, '\n')
    LOOP
      IF term <> '' THEN
        sensitive_terms_regexps := sensitive_terms_regexps || format('\m%s\M', term);
      END IF;
    END LOOP;

    found_match := FALSE;

    <<fields_loop>>
    FOREACH field IN ARRAY fields
    LOOP

      FOREACH term_regexp IN ARRAY sensitive_terms_regexps
      LOOP
        found_match := field ~* term_regexp;
        EXIT fields_loop WHEN found_match;
      END LOOP;

    END LOOP;

    RETURN found_match;
  END;
$$
LANGUAGE plpgsql;
```

</details>

I bet Perl might be the better option for this. In any case, it's at least _possible_ with a function of this sort, even if there are problems we need to account for and fix.

Of course, there's also the glaring issue of parsing the terms and building the regex list on every call to the function. I'm going to guess the "right" way to do this would be to have a table of the regex patterns that the function queries from or is required in the argument list.

Maybe we can discuss this tomorrow during the team meeting and come up with a decision of whether we feel this is sufficiently doable that we can move forward with this as the plan, instead of doing it in the ingestion worker. I feel confident we can come up with _something_ that will work, even if we run into some snags.

Reply:
@stacimc and I spent about 90 minutes or so today talking about this. We sorted out what seemed to be all the broad and many of the specific details required if we put the sensitive text detection into ingestion rather than in the data refresh.

Ultimately, while we came up with an idea that would work, it was _dramatically_ more complicated than this approach, and especially became complicated when trying to consider and account for the various race conditions that could occur with ingestion DAGs. It was all plausible, by our measure, but much more complex than the indexer worker version.

Therefore, (Staci, please correct me if I am wrong in this conclusion) we will stick with the indexer worker based approach as discussed in the current version of the IP, with the modifications I mentioned [in this comment](https://github.com/WordPress/openverse/pull/4904#issuecomment-2351839786).

I will also update the IP to include an explicit mention of this discussion and, without too much detail (there is a lot), note that the complexity outweighed any conceivable benefits we came up with, of which all were future theoretical benefits, rather than immediate ones.

Reply:
Details for this are in this commit: https://github.com/WordPress/openverse/pull/4904/commits/27b3781775231005ce3eddd6cff2c5d21afbdad8